### DISCUSSION TITLE
simulating runoff

### DISCUSSION THREAD

**Discussion by nlosacco** [idaholab/moose#25591] [Q&A Modules: Porous Flow]
Hi all,

I am trying to simulate runoff on a slope, i.e. the condition at which rainfall stops infiltrating due to saturation of the soil at the surface and flows as a thin water film downslope. This condition can be assumed to be represented by u=0 with u the porewater pressure at the soil surface.
So, using the PorousFlowSink to simulate prescribed variable rainfall infiltration in an initially unsaturated slope, so with initial u<0, when u=0 is reachedI'd like to be able to automatically switch to a flow which enforces the condition u=0 at the slope surface.
Does a feature like that exist?

**Comment by GiudGiud** [comment]
@cpgr

**Reply by cpgr** [reply]
I think that you can use `PorousFlowPiecewiseLinearSink` to do this - see https://mooseframework.inl.gov/modules/porous_flow/boundaries.html

**Reply by nlosacco** [reply]
Hi,

thanks for your reply. I already went through that page but I don’t think it is relevant to my case. I want to apply a flux independently of the pressure u value. But I want to make sure that u > 0 is never reached! So if during the calculations a value of u > 0 is predicted as a consequence of the prescribed flux, the latter should be automatically changed so that u=0 is achieved.

**Reply by nlosacco** [reply]
As a mechanical analogy you can think of a horizontal (nonlinear) cantilever beam with vertical point load prescribed at the tip and a unilateral contact constraint limiting the maximum deflection. Regardless of the applied load value, when the max deflection is reached no further vertical displacement of the tip is allowed, until the load is duly reduced

**Reply by cpgr** [reply]
Hmm, could you use a parsed `flux_function` in the sink to make the flux 0 when u > 0? Otherwise you might have to look at the conditional controls system to see if there is something that you can use there. @WilkAndy would probably know this, but is on leave until next week.

**Reply by nlosacco** [reply]
Thanks @cpgr , I’ll give a try. I am just afraid that those solution, parsed flux function and conditional controls, work in an explicit way, i.e. the check u>0 is done at the beginning of a step and nothing prevents significant overshooting to occur in the previous step if the flux is large enough… I’ll try and also wait for @WilkAndy suggestions

**Reply by nlosacco** [reply]
Hi all,

@cpgr , @WilkAndy , I had to put this aside for a while but now I'm on it again. Wondering if you have any suggestions on how to simulate the runoff. I think the way to go is imposing a time-varying flux through a `PorousFlowSink`, using a `ConditionalFunctionEnable` to check wether the porewaterpressure u is close to zero, then switching to a `PenaltyDirichletBC` to impose u=0 through a purposely calculated flow, if the latter condition is met.

About this I have a more general question which I am posing in https://github.com/idaholab/moose/discussions/26203

**Reply by arndflatten** [reply]
I think you don't need to switch the boundary condition from a fluxBC-type like PorousFlowSink (that you use for the unsaturated case where rain can still infiltrate into the model) to a dirichletBC (in case the pressure exceeds u>0). You may still want to use the fluxBC in the case of u>0 to prescribe a flux that automatically enforces the fluid pressure to be u=0. Try the [PorousFlowPiecewiseLinearSink](https://mooseframework.inl.gov/source/bcs/PorousFlowPiecewiseLinearSink.html) with your rainfall condition for u<0 and the sink condition with gradient being 1/L for u>0, see section Numerical Implementation in https://mooseframework.inl.gov/modules/porous_flow/boundaries.html.
